src/driver/kvm.rs

```diff
@@ -474,6 +455,1009 @@ mod tests {
+            .times(1)
+            .returning(move |_| Ok(access));
+
+        let mut kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+        let paddr: u64 = 0;
+        let result = Kvm::get_page_access(&mut kvm, paddr);
+
+        assert!(result.is_ok(), "Expected ok, got error instead!");
+    }
+
+    #[test]
+    fn test_get_page_access_fails_if_kvmi_get_page_access_fails() {
+        let mut kvmi_mock = MockKVMi::default();
+        setup_kvmi_mock(&mut kvmi_mock);
+        kvmi_mock.expect_get_page_access().times(1).returning(|_| {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "something went wrong",
+            ))
+        });
+
+        let mut kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+        let paddr: u64 = 0;
+        let result = Kvm::get_page_access(&mut kvm, paddr);
+
+        assert!(result.is_err(), "Expected error, got ok instead!");
+    }
+
+    #[test]
+    fn test_get_max_physical_addr_succeeds() {
+        let mut kvmi_mock = MockKVMi::default();
+        setup_kvmi_mock(&mut kvmi_mock);
+        kvmi_mock
+            .expect_get_maximum_gfn()
+            .times(1)
+            .returning(move || Ok(0));
+
+        let mut kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+        let result = Kvm::get_max_physical_addr(&mut kvm);
+
+        assert!(result.is_ok(), "Expected ok, got error instead!");
+    }
+
+    #[test]
+    fn test_get_max_physical_addr_fails_if_kvmi_get_maximum_gfn_fails() {
+        let mut kvmi_mock = MockKVMi::default();
+        setup_kvmi_mock(&mut kvmi_mock);
+        kvmi_mock.expect_get_maximum_gfn().times(1).returning(|| {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "something went wrong",
+            ))
+        });
+
+        let mut kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+        let result = Kvm::get_max_physical_addr(&mut kvm);
+
+        assert!(result.is_err(), "Expected error, got ok instead!");
+    }
+
+    #[test]
+    fn test_resume_vcpu_with_multiple_pause_events() {
+        let mut kvmi_mock = MockKVMi::default();
+        setup_kvmi_mock(&mut kvmi_mock);
+        kvmi_mock
+            .expect_wait_and_pop_event()
+            .times(5)
+            .returning(|_| {
+                Ok(Some(KVMiEvent {
+                    vcpu: 0,
```

Comment:
Returning 5 pause events for the same vcpu? This is unlikely to happen. You could pass in an outer varable which you increment every time you return something here.